nixos/modules/services/databases/postgresql.nix

```diff
@@ -631,6 +744,48 @@ in
         unitConfig.RequiresMountsFor = "${cfg.dataDir}";
       };
 
+      system.activationScripts.detect-previous-postgresql-installation = lib.mkIf (cfg.upgrade.enable && cfg.upgrade.enablePreviousInstallationAutodetection) {
+        deps = [ "etc" ];
+        text = ''
+          previousPgDetect() {
+            echo "Trying to detect prevoius PostgreSQL installation, because 'current' symlink does not exist in 'config.services.postgresql.databaseDir'."
+
+            if [[ ! -x /run/current-system/sw/bin/systemctl ]]; then
+              echo "systemctl binary is missing or is not executable. This is ok on fresh deployments. Not running previous PostgreSQL installation detection."
+              return
+            fi
+
+            prev_postgresql_env="$(/run/current-system/sw/bin/systemctl show postgresql.service --property=Environment --value || true)"
+            if [[ -z "$prev_postgresql_env" ]]; then
+              echo "Cannot load old PostgreSQL Environment from systemctl."
+              return
+            fi
+            old_data_dir="$(export $prev_postgresql_env; echo "''${PGDATA:-}")"
+            if [[ -z "$old_data_dir" ]]; then
+              echo "Cannot detect old PostgreSQL data dir!"
+              return
+            fi
+            old_bin=$(export $prev_postgresql_env; command -v postgres)
+            if [[ -z "$old_bin" ]]; then
+              echo "Cannot detect old PostgreSQL binary!"
+              return
+            fi
```

Comment:
The idea was to specify only the "final state" and let the system figure out how to do the upgrade.
This was one of the main points in the original PR discussion.

Also this is only one-time script, when the 'current' symlink does not exist and will never run again. And as mentiond above, I expect for most people it will never run.

e.g. we are upgrading from an unsupported version, which we cannot reference from the nixpkgs